Foxtrot/Driver/Rewriting/PostRewriteChecker.cs

```diff
@@ -0,0 +1,169 @@
+﻿// CodeContracts
+// Copyright (c) Microsoft Corporation
+// All rights reserved. 
+// MIT License
+// Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated documentation files (the "Software"), to deal in the Software without restriction, including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and to permit persons to whom the Software is furnished to do so, subject to the following conditions:
+// The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.
+// THE SOFTWARE IS PROVIDED *AS IS*, WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+
+namespace Microsoft.Contracts.Foxtrot
+{
+    using System;
+    using System.CodeDom.Compiler;
+    using System.Collections.Generic;
+    using System.Compiler;
+    using System.Diagnostics.CodeAnalysis;
+    using System.Diagnostics.Contracts;
+    using Microsoft.Contracts.Foxtrot.Properties;
+
+    /// <summary>
+    ///     A <see langword="sealed"/> class that reports error in case of any occurence of call to Code Contracts' methods.
+    /// </summary>
+    /// <remarks>
+    ///     Functionality provided by this class is being used in Foxtrot (ccrewrite) to ensure that all calls
+    ///     to Code Contracts' methods were rewritten. Otherwise "An assembly must be rewritten..." may be reported in
+    ///     in case of rewriter failure.
+    /// </remarks>
+    public sealed class PostRewriteChecker : Inspector
+    {
+        /// <summary>
+        ///     Holds a read-only reference to a stack which tracks visited members.
+        /// </summary>
+        private readonly Stack<Member> visitedMembers;
+
+        /// <summary>
+        ///     Holds a read-only reference to an error handler.
+        /// </summary>
+        private readonly Action<CompilerError> handleError;
+
+        /// <summary>
+        ///     Initializes a new instance of the <see cref="PostRewriteChecker"/> class with
+        ///     a reference to an action responsible for handling reported errors.
+        /// </summary>
+        /// <param name="handleError">
+        ///     An action that will be called for each error occurred.
+        /// </param>
+        public PostRewriteChecker(Action<CompilerError> handleError)
+        {
+            Contract.Requires(handleError != null);
+            Contract.Ensures(object.ReferenceEquals(this.handleError, handleError));
```

Comment:
I would expect that this kind of relationship can be inferred by static checker. However in some cases I was forced to add Ensures like this (using equality operator not ReferenceEquals) to mute cccheck warnings. This is more like a habit, which may be based on some early versions of CodeContracts.